DM-22964: Software packaging for deployment

mxk62 opened the pull request.

mxk62 committed:
Include LSST license

Added copyright and licensing information.

mxk62 committed:
Add EUPS table

mxk62 committed:
Provide support for LSST build system

Restructured project directory layout and added files required by LSST
build system, scons.

mxk62 committed:
Move application logic to main()

mxk62 committed:
Improve docstrings

Added few missing doctrings and fixed existing ones which were outdated
or incomplete.

mxk62 committed:
Rewrite imports using __all__

mxk62 committed:
Add a directory tree for future documentation

Added a directory tree for future documentation structured as described
in DM Developer Guide.

mxk62 committed:
Rename package

Renamed package to accommodate future development of the code running on
an endpoint site.

mxk62 committed:
Disable tests requiring a running ssh daemon

Few of the unit tests require a running ssh daemon to pass.  Setting up
a ssh daemon in a container just for these tests seems a bit excessive.
Hence I temporarily disabled them until I figure out how to rewrite them
in a way which does not need a running sshd.

mxk62 committed:
Add some logging to move command

Added some logging to Mover class.

mxk62 committed:
Replace os.rename() with shutil.move()

Function os.rename() only works if source and destination are on the
same device which causes problems with GPFS.  Replaced it with
shutil.move() which is supposed to be more robust.

mxk62 committed:
Call Wiper() unconditionally after transfer

Porter threads populate a queue with the names of the successfully
transferred files.  The size of the queue was used to determine if the
command removing possible empty directories on the endpoint site should
be run.  However, the same queue is being drained by the background
thread moving files from the buffer to the holding area.  It renders
size of the queue rather poor indicator if any files were transferred.
Hence, I modified to code to always call the clean-up command on the
endpoint site after all files are processed.

mxk62 committed:
Allow to configure port for SSH connenctions

All command using ssh/scp were using default port 22 for connection
without possibility to change it (e.g. for tests requiring ssh tunneling
using unprivileged ports).  Made changes allowing to specify port to use
in the configuration file.

mxk62 committed:
Separate storage area from the butler repository

So far the Butler repository served also as a permanent storage area
(i.e. they were no other copies of the images besides the repo after
successful ingestion).  However, due to the Butler being in a state of
flux the need to re-create a repository from scratch arises quite often.
Thus to make it easier I changed the buffer manager's order of
operation.  Now:

1. The buffer manager transfers a file to endpoint's staging area,
2. After the transfer is complete, it moves that file to a permanent
   storage area.
3. Then it makes a hard link to that file in endpoint's buffer.

which effectively separates storage area from the butler repository.

mxk62 committed:
Add a quickstart guide

Only API documentation was available so far.  Added a quickstart guide
describing how to install and use DBB buffer manager.

MichelleGower commented on bin.src/transd.py:
```diff
@@ -81,7 +102,9 @@ def background_thread(cmd, pause=1):
 
 
-if __name__ == "__main__":
+def main():
```
In DM-22677, besides turning the daf_butler's drivers into functions, the functions were moved to the python directory and unit tests written.   I'm having problems finding the driver behind the changes (LSST developer guide, RFC, etc).

mxk62 committed:
Move main() to a separate module

Following DM-22677, I moved main() and all auxiliary function to a
separate module.

mxk62 committed:
Remove credentials from config file
